js/tests/unit/dropdown.spec.js

```diff
@@ -367,6 +367,48 @@ describe('Dropdown', () => {
       dropdown.toggle()
     })
 
+    it('should toggle a dropdown with a only a valid virtual element reference', done => {
+      fixtureEl.innerHTML = [
+        '<div class="dropdown">',
+        '  <button class="btn dropdown-toggle visually-hidden" data-bs-toggle="dropdown" aria-expanded="false">Dropdown</button>',
+        '  <div class="dropdown-menu">',
+        '    <a class="dropdown-item" href="#">Secondary link</a>',
+        '  </div>',
+        '</div>'
+      ].join('')
+
+      const btnDropdown = fixtureEl.querySelector('[data-bs-toggle="dropdown"]')
+      const dropdownEl = fixtureEl.querySelector('.dropdown')
+      expect(() => new Dropdown(btnDropdown, {
+        reference: {}
+      })).toThrow()
+      expect(() => new Dropdown(btnDropdown, {
+        reference: {
+          getBoundingClientRect: 'not-a-function'
+        }
+      })).toThrow()
+      const dropdown = new Dropdown(btnDropdown, {
+        reference: {
+          getBoundingClientRect: () => ({
+            width: 0,
+            height: 0,
+            top: 0,
+            right: 0,
+            bottom: 0,
+            left: 0
+          })
```

Comment:
It would be better if you create a spy and then expect that to have been called after showing the dropdown.

Reply:
I was only testing the proper handling of types in general but adding a spy is a good idea. Will update this soon.

Reply:
This turned out to be a bit tricky as the internal `update` method on popper elements in asynchronous and popper does not expose events as bootstrap does. Luckily however there is a callback for the first update which I used instead.